src/java.base/share/classes/java/lang/classfile/Signature.java

```diff
@@ -185,88 +185,99 @@ public static ClassTypeSig of(ClassTypeSig outerType, String className, TypeArg.
     /**
      * Models the type argument.
      *
+     * @sealedGraph
      * @since 22
      */
     @PreviewFeature(feature = PreviewFeature.Feature.CLASSFILE_API)
-    public sealed interface TypeArg
-            permits SignaturesImpl.TypeArgImpl {
+    public sealed interface TypeArg {
 
         /**
-         * Indicator for whether a wildcard has default bound, no bound,
-         * an upper bound, or a lower bound
-         *
-         * @since 22
+         * Models an unbounded type argument {@code *}.
+         * @since 23
          */
         @PreviewFeature(feature = PreviewFeature.Feature.CLASSFILE_API)
-        public enum WildcardIndicator {
-
-            /**
-             * default bound wildcard (empty)
-             */
-            DEFAULT,
-
-            /**
-             * unbounded indicator {@code *}
-             */
-            UNBOUNDED,
+        public sealed interface Unbounded extends TypeArg permits SignaturesImpl.UnboundedTypeArgImpl {
+        }
 
-            /**
-             * upper-bounded indicator {@code +}
-             */
-            EXTENDS,
+        /**
+         * Models a type argument with an explicit bound type.
+         * @since 23
+         */
+        @PreviewFeature(feature = PreviewFeature.Feature.CLASSFILE_API)
+        public sealed interface Bounded extends TypeArg permits SignaturesImpl.TypeArgImpl {
 
             /**
-             * lower-bounded indicator {@code -}
+             * Models a type argument's wildcard indicator.
+             * @since 23
              */
-            SUPER;
+            @PreviewFeature(feature = PreviewFeature.Feature.CLASSFILE_API)
+            public enum WildcardIndicator {
+
+                /**
+                 * no wildcard (empty), an exact type
+                 */
+                DEFAULT,
```

Comment:
I’m personally in favour of naming this `NONE`.

Reply:
A belated +1 to this, at least not calling it `DEFAULT` because that has at best no meaning, and at worst it's outright confusing. If not `NONE`, then what about `EXACT`?

In my own old parsing code I have an enum called `Variance` with values `INVARIANT`, `COVARIANT`, and `CONTRAVARIANT`, but that might be a bit confusing considering that `EXTENDS` and `SUPER` match the Java language keywords directly.

As an aside, the javadoc here could be a little better IMO.

Reply:
Performed the rename. `NONE` is better than `EXACT`, as there indeed is no indicator char for this bounded type argument.

I have added the terms of `invariant` `covariant` `contravariant` to the docs, so that they can be searched with the Javadoc search function. (These are official CS terms so shouldn't be ambiguous)

I decided not too add exact descriptions, like "method parameters EXTENDS can only accept null" or "method return type SUPER can only return Object". These details of Java type system may be subject to change in the future, and this is really describing details of a ClassFile attribute instead of the Java type system.